File: PHP/Car/Car.php
<?php
    require_once 'Point.php';
    require_once 'Exceptions/OutOfFuel.php';
    require_once 'Exceptions/ToMuchFuel.php';
    
    class Car {
        private $model;
        private $location;
        private $fuelConsumption;
        private $fuelAmount, $fuelCapacity;
        
        private function isNumber($value) {
            if ( is_numeric($value) && $value > 0 ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        private function isPoint($value) {
            if ( $value instanceof Point ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        private function isString($value) {
            if ( is_string($value) ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        private function validate($value) {
            if ( is_numeric($value) ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        public function __construct($capacity = 60, $consumption = 0.6, $location = null, $model = 'Mercedes') {
            $this->fuelCapacity = $this->isNumber($capacity);
            $this->fuelConsumption = $this->isNumber($consumption);
            $this->location = is_null($location) ? new Point(0, 0) : $this->isPoint($location);
            $this->model = $this->isString($model);
            $this->fuelAmount = 0;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new InvalidAttribute("Attribute error: attribute '$name' not found.");
        }
        
        public function drive(Point $destination) {
            $neededFuel = $this->location->distance($destination) * $this->fuelConsumption;
    
            if ( $this->fuelAmount == 0 ) {
                throw new OutOfFuel("Your fuel tank is empty! Needed fuel for travel is $neededFuel liters.");
            }
            if ( $this->fuelAmount < $neededFuel ) {
                echo "Can't move! Needed fuel for travel is $neededFuel liters.";
                return;
            }
        
            $this->fuelAmount -= $neededFuel;
            $this->fuelCapacity -= $this->fuelAmount;
            $this->location = $destination;
        }

        public function travel($x, $y) {
            $this->validate($x);
            $this->validate($y);
            $this->drive(new Point($x, $y));
        }

        public function refill($fuel) {
            $this->isNumber($fuel);
                        
            if ( $this->fuelAmount + $fuel > $this->fuelCapacity ) {
                $this->fuelAmount = $this->fuelCapacity;
                throw new ToMuchFuel("To much fuel! Your fuel tank is full.");
            }   
            $this->fuelAmount += $fuel;
        }
    
        public function __toString() {
            return "Characteristics of the car:" . PHP_EOL
                . "Car model - $this->model." . PHP_EOL
                . "Car location - $this->location." . PHP_EOL
                . "Fuel amount - $this->fuelAmount liters." . PHP_EOL
                . "Tank capacity - $this->fuelCapacity liters." . PHP_EOL
                . "Fuel consumption - $this->fuelConsumption liters.";
        }
    }
?>

File: W5/Car/Car.py
from Point import Point
from Exceptions import OutOfFuelError, ToMuchFuelError, FuelCapacityError, FuelConsumptionError


class Car:
    def __init__(self, capacity: float = 60, consumption: float = 0.6,
                 location: 'Point' = Point(0.0, 0.0), model: str = 'Mercedes') -> None:
        self._fuel_amount = 0

        if capacity <= 0:
            raise FuelCapacityError('You entered incorrect tank capacity!')
        self._fuel_capacity = float(capacity)

        if consumption <= 0:
            raise FuelConsumptionError('You entered incorrect fuel consumption!')
        self._fuel_consumption = float(consumption)

        if not isinstance(location, Point):
            raise TypeError()
        self._location = location

        self._model = str(model)

    def __str__(self) -> str:
        return f'Characteristics of the car:\n' \
               f'The car model {self._model}.\n' \
               f'The car location {self._location}.\n' \
               f'Fuel amount {self._fuel_amount}.\n' \
               f'Tank capacity {self._fuel_capacity}.\n' \
               f'Fuel consumption {self._fuel_consumption}.'

    def __repr__(self) -> str:
        return f'Characteristics of the car:\n' \
               f'The car model {self._model}.\n' \
               f'The car location {self._location}.\n' \
               f'Fuel amount {self._fuel_amount}.\n' \
               f'Tank capacity {self._fuel_capacity}.\n' \
               f'Fuel consumption {self._fuel_consumption}.'

    @property
    def fuel_amount(self) -> float:
        return self._fuel_amount

    @property
    def fuel_capacity(self) -> float:
        return self._fuel_capacity

    @property
    def fuel_consumption(self) -> float:
        return self._fuel_consumption

    @property
    def location(self) -> 'Point':
        return self._location

    @property
    def model(self) -> str:
        return self._model

    def drive_point(self, destination: 'Point') -> None:
        if not isinstance(destination, Point):
            raise TypeError()

        needed_fuel = self._location.distance(destination) * self._fuel_consumption

        if self._fuel_amount == 0:
            raise OutOfFuelError(f'Your fuel tank is empty! Needed fuel for travel is {needed_fuel} liters.')
        if self._fuel_amount < needed_fuel:
            print(f"Can't move! Needed fuel for travel is {needed_fuel} liters.")
            return

        self._fuel_amount -= needed_fuel
        self._fuel_capacity -= self._fuel_amount
        self._location = destination

    def drive_coord(self, x: float, y: float) -> None:
        self.drive_point(Point(x, y))

    def refill(self, fuel: float) -> None:
        if self._fuel_amount + fuel > self._fuel_capacity:
            self._fuel_amount = self._fuel_capacity
            raise ToMuchFuelError('To much fuel! Your fuel tank is full.')
        elif fuel < 0:
            self._fuel_amount = 0
            return

        self._fuel_amount += fuel
